### PR TITLE
martin/qa-challenge

### PR DESCRIPTION
**Pull request by Martin-Lozdan-Parra**
## Description

Created tests for logIn, logOut and signUp flows, including API validations. Also included CI execution manually and on every PR.

Please include a summary of the changes and the related issue. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: cypress/pages/SignUpPage.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        cy.get(this.city).type(addressInfo.city);
+        cy.get(this.zipcode).type(addressInfo.zipcode);
+        cy.get(this.mobileNumber).type(addressInfo.mobileNumber);
+    }
+
+    
+    // Create account by clicking the create account button
+    
+    clickCreateAccountButton(): void {
+        cy.get(this.createAccountButton).click();
```
Comment (ulises-jeremias) [inline cypress/pages/SignUpPage.ts:128]: **Test Reliability & Wait Strategies**

Consider adding explicit waits for better stability

```ts
cy.get(this.createAccountButton)
  .should('be.visible')
  .and('be.enabled')
  .click();
```

File: cypress/pages/NavBar.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+﻿import ProductsPage from "./ProductsPage";
+import CartPage from "./CartPage"
+
+export default class NavBar {
+    private readonly header = '#header';
```
Comment (ulises-jeremias) [inline cypress/pages/NavBar.ts:5]: **Selector Strategy Consistency**
* Your use of `data-qa` attributes is excellent! Consider standardizing this approach throughout
* Mix of `data-qa` (robust) and generic selectors like `#header` (fragile) could be unified
* This will make tests more maintainable and resilient to UI changes

File: cypress/e2e/logIn_user.cy.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+describe('Login User wrong password', () => {
+
+    let userData: User;
+    
+      before(() => {
+        // Generate random user data for signup
+        userData = TestDataGenerator.generateUserData();
+    
+        // Create the user through API
+        createUser(userData);
```
Comment (ulises-jeremias) [inline cypress/e2e/logIn_user.cy.ts:49]: **Error Handling & Recovery**
```typescript
// Consider adding error handling for API operations
try {
  createUser(userData);
} catch (error) {
  cy.log('User creation failed, attempting cleanup...');
  // Handle conflicts or retry logic
}
```

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+This repository also has the test triggered on every pull request.
+Its defaulted to run all specs files, to execute a specific, use the path of the spec files as parameter
+cypress/e2e/logOut_user.cy.ts
+
+## Improvements
+
+Fields data of signUp forms accepts any character, even emojis on the address, which is a major problem if we think on products deliveries. Having field validations and creating negative tests cases like Wrong adress, wrong phone... etc would be a nice improvement
+
+a
```
Comment (ulises-jeremias) [inline docs/README.md:44]: **Documentation Enhancement**
* The `docs/README.md` needs formatting fixes and more comprehensive content
* Consider adding troubleshooting guide, test strategy explanation, and setup prerequisites
* Main README still contains the challenge template — replace with project-specific documentation

File: cypress/pages/LoginPage.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    verifySignupErrorMessage(message: string): void {
+        cy.get(this.signupForm)
+            .contains('Email Address already exist!').should('be.visible');
+
+    }
+
+    verifyEnterAccountMessageIsVisible() {
+        cy.contains(/ENTER ACCOUNT INFORMATION/i).should('be.visible');
+    }
+} export default new LoginPage();
```
Comment (ulises-jeremias) [inline cypress/pages/LoginPage.ts:62]: Add page load verification method

```typescript
waitForPageLoad(): void {
  cy.get(this.loginForm).should('be.visible');
  cy.get(this.signupForm).should('be.visible');
}
```

File: cypress/utils/Helper.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      company: user.company,
+      address1: user.address1,
+      address2: user.address2,
+      country: user.country,
+      zipcode: user.zipcode,
+      state: user.state,
+      city: user.city,
+      mobile_number: user.mobileNumber,
+    },
+    failOnStatusCode: true,
```
Comment (ulises-jeremias) [inline cypress/utils/Helper.ts:27]: Add error handling for API calls

```ts
export function createUser(user: User) {
  return cy.request({
    method: 'POST',
    url: 'https://automationexercise.com/api/createAccount',
    form: true,
    body: { /* user data */ },
    failOnStatusCode: false, // Handle errors gracefully
  }).then((response) => {
    if (response.status !== 200) {
      throw new Error(`User creation failed: ${response.body}`);
    }
    return response;
  });
}
```

File: cypress.config.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    setupNodeEvents(on, config) {
+      // implement node event listeners here
+      return config;
+    },
+  },
+  retries: {
+    runMode: 2,
+    openMode: 0,
+  },
+  env: {
```
Comment (ulises-jeremias) [inline cypress.config.ts:24]: Consider adding environment-specific configs

```ts
env: {
  apiUrl: 'https://automationexercise.com/api',
  retryAttempts: 3,
  testTimeout: 10000,
}
```

**Review by ulises-jeremias — COMMENTED** [review]
**Hey @Martin-Lozdan-Parra  ! 👋**

Thank you for this solid submission! 🙌 Your test automation suite demonstrates a strong foundation in modern testing practices and shows clear understanding of Cypress, TypeScript, and test automation principles.

At NaNLABS, our goal is that everyone walks away with **something valuable to explore or reflect on**, so we'll share some observations and suggestions that could help take your automation skills to the next level. These are meant as growth opportunities rather than criticisms 🚀

---

## 🌟 Highlights

Here are some of the things we really appreciated:

* ✅ **Excellent Page Object Model implementation** — clean inheritance with BasePage and good separation of concerns
* ✅ **Smart hybrid UI + API approach** — using API calls for setup/teardown makes tests faster and more reliable
* ✅ **Strong TypeScript usage** — proper interfaces, type safety, and modern ES6+ patterns
* ✅ **Good test data strategy** — thoughtful use of Faker.js for dynamic data generation
* ✅ **Modular architecture** — well-organized folder structure with clear responsibilities
* ✅ **CI/CD integration** — GitHub Actions workflow properly configured for automated testing
* ✅ **Multiple test scenarios** — covers happy path, negative cases, and edge cases

---

## 💡 Suggestions for Growth

Your foundation is solid! Here are some areas where you could take it even further:

### 🎯 **Test Reliability & Wait Strategies**
```typescript
// Current approach in SignUpPage.ts
cy.get(this.createAccountButton).click();

// Consider adding explicit waits for better stability
cy.get(this.createAccountButton)
  .should('be.visible')
  .and('be.enabled')
  .click();
```

### 🔍 **Selector Strategy Consistency**
* Your use of `data-qa` attributes is excellent! Consider standardizing this approach throughout
* Mix of `data-qa` (robust) and generic selectors like `#header` (fragile) could be unified
* This will make tests more maintainable and resilient to UI changes

### 📊 **Enhanced Test Coverage**
* **Form validation testing** — empty fields, invalid email formats, special characters
* **Mobile/responsive testing** — different viewport sizes and touch interactions  
* **Error handling scenarios** — network failures, API timeouts, server errors
* **Accessibility testing** — keyboard navigation, screen reader compatibility

### 🛡️ **Error Handling & Recovery**
```typescript
// Consider adding error handling for API operations
try {
  createUser(userData);
} catch (error) {
  cy.log('User creation failed, attempting cleanup...');
  // Handle conflicts or retry logic
}
```

### 📚 **Documentation Enhancement**
* The `docs/README.md` needs formatting fixes and more comprehensive content
* Consider adding troubleshooting guide, test strategy explanation, and setup prerequisites
* Main README still contains the challenge template — replace with project-specific documentation

### 🧪 **Test Data Management Evolution**
* Current approach generates completely random data — great for avoiding conflicts!
* Consider adding deterministic test data for specific scenarios to aid debugging
* Implement test data cleanup strategies for better isolation

---

## 🎯 **Specific Technical Improvements**

### **In LoginPage.ts (lines 30-35):**
```typescript
// Add page load verification method
waitForPageLoad(): void {
  cy.get(this.loginForm).should('be.visible');
  cy.get(this.signupForm).should('be.visible');
}
```

### **In Helper.ts (lines 15-25):**
```typescript
// Add error handling for API calls
export function createUser(user: User) {
  return cy.request({
    method: 'POST',
    url: 'https://automationexercise.com/api/createAccount',
    form: true,
    body: { /* user data */ },
    failOnStatusCode: false, // Handle errors gracefully
  }).then((response) => {
    if (response.status !== 200) {
      throw new Error(`User creation failed: ${response.body}`);
    }
    return response;
  });
}
```

### **In cypress.config.ts:**
```typescript
// Consider adding environment-specific configs
env: {
  apiUrl: 'https://automationexercise.com/api',
  retryAttempts: 3,
  testTimeout: 10000,
}
```

---

## 🚀 **Advanced Opportunities**

If you want to explore further:

* 🔄 **Custom Cypress commands** — create reusable commands for common workflows
* 📈 **Test reporting** — integrate with tools like Mochawesome or Allure
* 🌐 **Cross-browser testing** — extend CI to run tests in Firefox, Safari, Edge
* 🔍 **Visual regression testing** — add screenshot comparison capabilities
* ⚡ **Performance testing** — integrate basic performance assertions
* 🐳 **Docker containerization** — make the test suite portable across environments

---

## ✅ Final Thoughts

This is a **well-structured and thoughtful submission** that demonstrates solid understanding of test automation principles. The hybrid UI/API approach shows practical experience, and the codebase is clean and maintainable.

Your technical skills are clearly strong, and with some focus on test reliability patterns and expanded coverage strategies, you'd be well-positioned to lead test automation initiatives in a production environment.

The areas we've highlighted are common challenges that even experienced automation engineers work to improve, so don't see them as shortcomings but as exciting opportunities to level up your expertise!

Thanks for the effort and care you put into this challenge — it's clear you understand what good test automation looks like 👏

---

*Keep building, keep learning, and happy testing! 🧪✨*
